DMCompiler/DM/Visitors/DMVisitorExpression.cs

```diff
@@ -434,56 +437,250 @@ public void VisitIdentifier(DMASTIdentifier identifier)
+                            if (property != null) {
+                                operation.Identifier = field;
+                                operation.GlobalId = null;
+                                operation.Path = property.Type;
+                            } else {
+                                var globalId = dmObject.GetGlobalVariableId(field);
+                                if (globalId != null) {
+                                    property = DMObjectTree.Globals[globalId.Value];
+
+                                    expr = new GlobalField(expr.Location, property.Type, globalId.Value);
+
+                                    var newOperationCount = operations.Length - i - 1;
+                                    if (newOperationCount == 0) {
+                                        Result = expr;
+                                        return;
+                                    }
+
+                                    operations = new Dereference.Operation[newOperationCount];
+                                    astOperationOffset = i + 1;
+                                    i = -1;
+                                }
+                            }
+
+                            if (property == null) {
+                                throw new CompileErrorException(deref.Location, $"Invalid property \"{field}\" on type {prevPath}");
+                            }
+
+                            if ((property.ValType & DMValueType.Unimplemented) == DMValueType.Unimplemented) {
+                                DMCompiler.UnimplementedWarning(deref.Location, $"{prevPath}.{field} is not implemented and will have unexpected behavior");
+                            }
+
+                            prevPath = property.Type;
+                            pathIsFuzzy = false;
+                        }
+                        break;
+
+                    case DMASTDereference.OperationKind.FieldSearch:
+                    case DMASTDereference.OperationKind.FieldSafeSearch:
+                        // TODO: im pretty sure types should be inferred if a field with their name only exists in a single place, sounds cursed though
+                        operation.Identifier = astOperation.Identifier.Identifier;
+                        operation.GlobalId = null;
+                        operation.Path = null;
+                        prevPath = null;
+                        pathIsFuzzy = true;
+                        break;
+
+                    case DMASTDereference.OperationKind.Index:
+                    case DMASTDereference.OperationKind.IndexSafe:
+                        // Passing the path here is cursed, but one of the tests seems to suggest we want that?
+                        operation.Index = DMExpression.Create(_dmObject, _proc, astOperation.Index, prevPath);
+                        operation.Path = null;
+
+                        // This is cursed too
+                        // prevPath = prevPath;
```

Comment:
What was this commented line for?

Reply:
I was expecting some weird fuckery with the inferred type of `L[x]` but I think I was wrong since it didn't hit any unit tests.